apps/explorer/src/components/ownedobjects/views/DynamicFieldsCard.tsx

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { useRpcClient } from '@mysten/core';
+import { useInfiniteQuery } from '@tanstack/react-query';
+import { useRef, useCallback, useEffect } from 'react';
+
+import { UnderlyingObjectCard } from './UnderlyingObjectCard';
+
+import { SyntaxHighlighter } from '~/components/SyntaxHighlighter';
+import { DisclosureBox } from '~/ui/DisclosureBox';
+import { ObjectLink } from '~/ui/InternalLink';
+import { LoadingSpinner } from '~/ui/LoadingSpinner';
+import { Tab, TabGroup, TabList, TabPanel, TabPanels } from '~/ui/Tabs';
+
+const MAX_PAGE_SIZE = 20;
+
+export function DynamicFieldsCard({ id }: { id: string }) {
+    const rpc = useRpcClient();
+    const {
+        data,
+        isInitialLoading,
+        isFetchingNextPage,
+        hasNextPage,
+        fetchNextPage,
+    } = useInfiniteQuery(
+        ['dynamic-fields', id],
+        ({ pageParam = null }) =>
+            rpc.getDynamicFields({
+                parentId: id,
+                cursor: pageParam,
+                limit: MAX_PAGE_SIZE,
+            }),
+        {
+            enabled: !!id,
+            getNextPageParam: ({ nextCursor, hasNextPage }) =>
+                hasNextPage ? nextCursor : null,
+        }
+    );
+
+    const observerElem = useRef<HTMLDivElement | null>(null);
+
+    const handleObserver = useCallback(
+        (entries: IntersectionObserverEntry[]) => {
+            const [target] = entries;
+            if (target.isIntersecting && hasNextPage) {
+                fetchNextPage();
+            }
+        },
+        [fetchNextPage, hasNextPage]
+    );
+
+    useEffect(() => {
+        const element = observerElem.current;
+        const option = { threshold: 0 };
+        const observer = new IntersectionObserver(handleObserver, option);
+        if (!element) return;
+        observer.observe(element);
+        return () => observer.unobserve(element);
+    }, [fetchNextPage, hasNextPage, handleObserver]);
+
+    if (isInitialLoading) {
+        return (
+            <div className="mt-1 flex w-full justify-center">
+                <LoadingSpinner />
+            </div>
+        );
+    }
+
+    // show the dynamic fields tab if there are pages and the first page has data
+    const hasPages = !!data?.pages?.[0].data.length;
```

Comment:
It seems that we don't need the items per page so I was wondering if we could flatten the array and have in the end just a list of items. Something like `data?.pages?.flat(2)`